Packaging+license

ShantanuKodgirwar opened the pull request:
This PR does the following:

- Adds an MIT license to the repository as there was none before. 
- Packages the repository by adding a `pyproject.toml` file instead of `setup.py` file as per the recent [PEP 621](https://peps.python.org/pep-0621/) standard.
- minor changes to .gitignore
- deletes `setup.py` file
- Adding installation instructions under `README.md`

Additionally, the package is not yet on [PyPI](https://pypi.org/), I'd suggest publishing the current version 0.1.0 to pypi for easy installation and use in future. For example `pip install ptylab`. The `pyproject.toml` was generated with the python package manager (Poetry)[https://github.com/python-poetry/poetry]. A simple command `poetry build --publish` should allow that. Needless to say, someone would first need to create an account on PyPI first. 

@larsloetgering @dboonz @MaisieD Please take a note.

roflmaostc commented:
Hi,

Thanks for taking the effort.

But, the Python should be also under the same license as [this](https://github.com/PtyLab/PtyLab.jl/blob/main/LICENSE.md). 

So I guess this would ban it from uploading it to PyPI?

ShantanuKodgirwar commented:
@roflmaostc I mean I am not sure if there is a limitation for which licenses are allowed. I saw packages that use Academic Free License (AFL). If this license is already similar to that, I think should be okay.

roflmaostc commented:
Hi!

Can you please summarize what you commited so far?

I think we should not host it on other servers. Can you clean that?

Maybe some Python people can check @dboonz @MaisieD :)?

ShantanuKodgirwar commented:
Sure! I'll repeat some of the stuff from the first comment. This PR includes the following

- As per your last comment, copied the license you have in the Julia package here. I am not so sure if this should affect publishing on pypi. It should be alright. 
- Packages the repository by adding a `pyproject.toml` file instead of `setup.py` file as per the recent [PEP 621](https://peps.python.org/pep-0621/) standard.
- `cupy` is a gpu dependency which I kept as optional and can be simply installed with the flag, `black` and `ipykernel` are dev dependencies.
- Updated `README.md` that mainly illustrates how to install the package from source for and without these flags along with directly cloning and installing the package in a separate virtual environment for development.
- Updating `.gitignore` with some minor additions.

ShantanuKodgirwar commented:
> I think we should not host it on other servers. Can you clean that?

I didn't get you here.

ShantanuKodgirwar commented:
@roflmaostc @dboonz @MaisieD 

I would also remove the `environment.yml` file from the repository, therefore removing reliance on conda for package management/virtual environment creation. They are incredibly slow when resolving any dependency issues when installing the package. Highly inflexible in terms of allowing optional dependencies or allowing more Python versions (user can have any Python installed in the system), currently we allow any Python version from 3.9 to 3.11, an `environment.yml` with this flexibility would be even slower to install. 

As written in the readme, for development purposes, it is better to either rely on [virtualenv](https://virtualenv.pypa.io/en/latest/installation.html) for creating a virtual environment or better with [poetry](https://python-poetry.org/docs/) as the `pyproject.toml`, I added is based on poetry core. 

Moreover, we can also mention the stuff with poetry if necessary (perhaps in a separate `CONTRIBUTING.md` file). 

For me these should be the following next steps:
- removing `environment.yml`
- publishing to pypi with the current version 0.1.0
- Optionally adding a `CONTRIBUTING.md` file for more detailed installation instructions
- Perhaps improving `README.md` by potentially adding a small example.

I think these steps should improve the visibility of the package as well.

ShantanuKodgirwar commented:
Manual installation of `cupy` is needed as it is based on individual CUDA version
- Updating `pyproject.toml` to remove the automatic GPU installation field
- Updating `README.md` with new instructions for `cupy` installation.

ShantanuKodgirwar commented:
@dboonz @roflmaostc I think we can now merge this. There are no breaking changes here anyways, just packaging. We should also put this on pypi for ease for install.

roflmaostc commented:
Thanks for cleaning this!

I'm still not sure if pypi agrees with our license. So I would not upload it.

But someone else with more Python authority (such as @dboonz) should merge this :)

ShantanuKodgirwar commented:
@dboonz @roflmaostc Adding instructions to use `conda` as many people prefer to use it in general. It is also easier to install `cupy` with conda. I tested this with Python 3.10.13 (ideally any version >=3.9, <3.12 should also work).

roflmaostc commented:
Thanks for doing all this! And sorry that I do not want to merge, don't have the experience regarding Python :)

ShantanuKodgirwar commented:
No worries. I did this anyway as I was testing it for myself. :) 
I also wrote to @dboonz about this if he decides to review and merge. But nonetheless, no pressure :)
I also have another branch with some bug fixes, but yeah I never thought about sending a PR. I think it makes more sense if there is someone active from the python development team to review that and merge it.

roflmaostc commented:
Actually, we probably should bump the version number, right? Since this is breaking, should be 0.2.0

ShantanuKodgirwar commented:
I mean there is no official versioning involved as of yet for the current `ptylab` version. The old one from `setup.py` was 0.0 (also not standard versioning :D), but yeah, the current version 0.1.0 is new if this gets merged. However, if you think 0.2.0 makes sense. I can make this quick change in `pyproject.toml`

roflmaostc commented:
ping @dboonz @MaisieD

ShantanuKodgirwar commented:
New commit adds additional note in README about using [Poetry](https://python-poetry.org/) a python package manager when dependencies are to be modified or added. This is because the `pyproject.toml` file was made using Poetry as its build-system. Poetry also comes with it's own excellent dependency resolver which would make sure nothing breaks in future when modifying dependencies.

dboonz commented:
Hi @ShantanuKodgirwar and @roflmaostc ,

I'm sorry for the lack of response, I have a new job and I didn't manage so far. 

It looks like you have done some very nice packaging work! I just tried it out, and it seems to work like a charm. That's really a big help and I agree that it will make it a lot easier for people to use. So far installation was a pain.

I don't know a lot about packaging, if we get into trouble in the future, would you be able to help out? Also, most of the people that develop this are not "real" programmers, could you put a small note about how to add packages using Poetry?  Just so they have an idea what to do.

Did you check that the examples still run?

I am not sure about the licensing, I think we should ask @larsloetgering  to look at that, he put quite some thought into it when we developed this package. I'll do some more testing tomorrow but in principle I'm happy to merge it. Thank you for the work!

ShantanuKodgirwar commented:
Hi @dboonz,

No worries :) 

> I don't know a lot about packaging, if we get into trouble in the future, would you be able to help out? Also, most of the people that develop this are not "real" programmers, could you put a small note about how to add packages using Poetry? Just so they have an idea what to do.

Sure thing! I'll definitely help out! :) About adding info about Poetry, I can create a `CONTRIBUTING.md` file (often a common practice). 

> I am not sure about the licensing, I think we should ask @larsloetgering to look at that, he put quite some thought into it when we developed this package. 

Sure, whatever you and Lars decide. I mean for now, the license is the same as in the Julia version, which should also be okay.

> I'll do some more testing tomorrow but in principle I'm happy to merge it. Thank you for the work!

Sounds great! I'll add the `CONTRIBUTING.md` file (tomorrow maybe ) and you can then merge! :)

larsloetgering commented:
Hi all,

As long as the licensing file contains an explicit statement that phase
focus holds a portfolio of licenses commercializing ptychography, I am fine
with alternative ways of redistribution. However, the statement as
currently is in our license file must be in there, otherwise we could get
into legal trouble.

Best wishes,
Lars

Am Sa., 13. Jan. 2024 um 18:05 Uhr schrieb Shantanu Kodgirwar <
***@***.***>:

> Hi @dboonz <https://github.com/dboonz>,
>
> No worries :)
>
> I don't know a lot about packaging, if we get into trouble in the future,
> would you be able to help out? Also, most of the people that develop this
> are not "real" programmers, could you put a small note about how to add
> packages using Poetry? Just so they have an idea what to do.
>
> Sure thing! I'll definitely help out! :) About adding info about Poetry, I
> can create a CONTRIBUTING.md file (often a common practice).
>
> I am not sure about the licensing, I think we should ask @larsloetgering
> <https://github.com/larsloetgering> to look at that, he put quite some
> thought into it when we developed this package.
>
> Sure, whatever you and Lars decide. I mean for now, the license is the
> same as in the Julia version, which should also be okay.
>
> I'll do some more testing tomorrow but in principle I'm happy to merge it.
> Thank you for the work!
>
> Sounds great! I'll add the CONTRIBUTING.md file (tomorrow maybe ) and you
> can then merge! :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PtyLab/PtyLab.py/pull/48#issuecomment-1890594539>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVIVGFTGOHURJWUQVSBZNWDYOK5GBAVCNFSM6AAAAAA3ERPLPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOJQGU4TINJTHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ShantanuKodgirwar commented:
@larsloetgering Thanks for your comment. In that case, the current `LICENSE` should be okay!

ShantanuKodgirwar commented:
@dboonz @roflmaostc 

I added the info on Poetry under `README.md` and didn't create`CONTRIBUTING.md` file to avoid hassle for users.
Additionally, deleted the files that are now redundant. Those were `environment.yml` and `install_linux.sh`. There are still additional files `run_tests.sh` and `test_gpu_installation.py` that should also be perhaps deleted or moved to some "tests" directory I suppose. 

Btw I didn't test the examples under the directory "example_scripts", but I did test the notebook `jupyter_tutorials_tutorial_CPM_sim.ipynb` under "jupyter_tutorials" in the past.

ShantanuKodgirwar commented:
@dboonz Feel free to merge once you have tested.

dboonz commented:
Great, looks terrific. Thank you for the effort! I'm merging now.
